### ISSUE TITLE
Support for GHA's setup-python pre-release version specifiers

### ISSUE THREAD

**Issue by hynek** [wntrblm/nox#685]
### How would this feature be useful?

Currently, nox supports the setup-python notation for stable Python versions perfectly.

You can write something like:

```yaml
jobs:
  tests:
    name: nox on ${{ matrix.runs-on }} / ${{ matrix.python-version }}
    runs-on: ${{ matrix.runs-on }}
    strategy:
      matrix:
        python-version: ["pypy3.8", "3.8", "3.9", "3.10"]
        runs-on: [ubuntu-latest]

    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}
      - python -m pip install --upgrade wheel virtualenv nox

      - run: nox --python ${{ matrix.python-version }}
```

And it just works. It even works with `"3.11.0-beta - 3.11"` – although by accident.

But it doesn't work at all with the modern `~3.12.0-0` (that matches both prereleases as well as eventual stable versions), which means that one is forced to keep kind of a "lookup table" like in [Towncrier](https://github.com/twisted/towncrier/blob/ffcc117225504d7275a9662be421d1494c33e907/.github/workflows/ci.yml#L61-L75) which is a bit of a bummer.

### Describe the solution you'd like

My proposal now is that nox should accept all `python-version` syntaxes.

Another great option would be to be able to say `nox -p current` or something like that, that will use the Python version that is running nox. That would essentially implement tox-gh-actions.

### Describe alternatives you've considered

Well, the table thing I've used in Towncrier that is also necessary for

**Comment by mayeut** [comment]
@hynek, you can use:
```yaml
jobs:
  tests:
    name: nox on ${{ matrix.runs-on }} / ${{ matrix.python-version }}
    runs-on: ${{ matrix.runs-on }}
    strategy:
      matrix:
        python-version: ["pypy3.8", "3.8", "3.9", "3.10", "3.12"]
        runs-on: [ubuntu-latest]

    steps:
      - uses: actions/checkout@v3
      - uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}
          allow-prereleases: true
      - python -m pip install --upgrade wheel virtualenv nox

      - run: nox --python ${{ matrix.python-version }}
```

**Comment by hynek** [comment]
Wow that looks cool – does it have any weird side-effects when running with say 3.10 etc? Is it a new option or why did everyone use the other syntax ever since? cc @webknjaz

**Comment by hynek** [comment]
> Either way, for this case, it's necessary to have some way of identifying if a thing is nightly-ish in order to allow failures in unstable runtimes.

Seems like the best solution here would be if setup-python would set some kind of variable to indicate that?

**Comment by hugovk** [comment]
> Wow that looks cool – does it have any weird side-effects when running with say 3.10 etc? 

No weird side effects: with `3.10` and `allow-prereleases: true`, it installs the latest alpha/beta/RC/full release available: currently 3.10.11. It won't go back in time to a pre-release.

(This is different to Travis CI, where `-dev` was a nightly build, but `setup-python` uses actual releases. https://github.com/deadsnakes/action is good for nightly builds on GHA.)

> Is it a new option or why did everyone use the other syntax ever since?

`allow-prereleases: true` is new last month: https://github.com/actions/setup-python/releases/tag/v4.6.0

Some people don't like `3.12-dev` because they don't want to delete the `-dev` from `["3.8", "3.9", "3.10", "3.11", "3.12-dev"]` when the full release comes out in October, so `allow-prereleases: true` is neater for them, can keep `["3.8", "3.9", "3.10", "3.11", "3.12"]` for the full cycle. (I saw one project that appended `-dev` like `python-version: ${{ matrix.python-version }}-dev`, to keep the matrix cleaner.)

And it makes things neater when passing in a Python matrix version to things like tox/nox/random script.

Of course, if you want to mark a build as unstable to allow failures, you'll either need to use `-dev`, or `~`, or add your own variable, or...

> Seems like the best solution here would be if setup-python would set some kind of variable to indicate that?

Yeah, that sounds good.

> Not everyone needs/cares about this, though. I recall @hugovk mentioning that he doesn't see such approach as viable and was pushing for this "match anything found" feature in the upstream action.

Yep, I've rarely needed to allow failures for `-dev` builds. I'd rather know about a hard failure sooner, and ideally fix the code quickly, or if takes longer/needs an upstream report, we can always comment it out, or consider allowing failures at that point. (I might reconsider if allowed failures was better built into GHA, similar to Travis, but I generally prefer to just avoid this extra complexity.)

**Comment by hynek** [comment]
Thanks to `allow-prereleases: true` , I don't think there's anything to do for Nox.
